pkg/nuget/parse.go

```diff
@@ -0,0 +1,84 @@
+package nuget
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/aquasecurity/go-dep-parser/pkg/types"
+	"io"
+)
+
+// The lockfile that nuget uses (packages.lock.json) works basically like this:
+// Each 'build type' (for example .NETCoreApp,Version=v3.1) have a set of dependencies.
+// Those dependencies can be 'Direct', 'Transitive' and 'Project'
+// (meaning of those are quite self explanatory I think). The dependency consists of a string
+// value key (package name) and an object, from which we need to fetch a few different values.
+// The following values should be parsed right away: resolved (resolved version),
+// type (this information might be worth having) and finally, it's dependencies.
+// The dependencies in this stage of the file are only a string map where the key is package name
+// and value is package version.
+
+type LockFile struct {
+	Version      int
+	Dependencies map[string]map[string]Dependency
+}
+
+type Dependency struct {
+	Type         string            `json:"type"`
+	Resolved     string            `json:"resolved"`
+	Dependencies map[string]string `json:"dependencies"`
+	ContentHash  string            `json:"contentHash"`
+}
+
+func Parse(r io.Reader) ([]types.Library, error) {
+	var lockFile LockFile
+	decoder := json.NewDecoder(r)
+	err := decoder.Decode(&lockFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var libraries []types.Library
+	unique := map[string]struct{}{}
+
+	// The first loop is the targets (.NETCoreApp,Version=v3.1)
+	// that means that it's not really a package, but there are dependencies
+	// under said target.
+	for _, targetContent := range lockFile.Dependencies {
+		// Here comes the 'real' dependencies.
+		for topPkgName, topPkgContent := range targetContent {
+			// And inside the dependency, there are another level of dependencies.
+			// But that's actually it, I swear!
+			for pkgName, version := range topPkgContent.Dependencies {
+				symbol := fmt.Sprintf("%s@%s", pkgName, version)
+				if _, ok := unique[symbol]; ok {
+					continue
+				}
```

Comment:
What happens if we have two transitive dependencies but of different versions, which one gets picked up? The first in the JSON? Is that desired behaviour?

Reply:
I would extend that argument to any dependency type (transitive or otherwise). What happens in such a case?

Reply:
The symbol depends on both the package name and versions, so in a case where two versions exist as dependencies, it _should_ pick both up as different dependencies.  
I think that is the expected/intended way, right?

I will update with a testcase for this, and if I missunderstod the expected behavior, please let me know! :)

Reply:
Update: Added test case for this. If it's not the intended behavior, please let me know and I'll fix it right away! :)

Reply:
Thanks for adding the test case! This makes sense to me. If a project is using two different versions of dependencies somehow, I agree both should be taken into account. 

cc @knqyf263 would you have more insight into this?

Reply:
It sounds fine.

Reply:
I missed those commands!
I generated a lockfile in one of my projects and removed some of the dependencies just to get a shorter list, but I'll create a new one via a similar command as the above!

Reply:
It makes sense. The lock file looks broken since you removed some dependencies to make it shorter. It would be great if you write commands down to reproduce the lock files for testing.

Reply:
While fixing this, I re-read the whole nuget package resolving docs to check why I got it wrong. This made me rewrite the algorithm a bit...
NuGet will use the version of the package "closest" to the application, that is, it will resolve to use a package that is a direct dependency above using a sub-dependency. So if a package is installed by the user, that is the one that it will use.

What I changed was the following:

Instead of looping through the sub packages at the same time as the top level packages, I changed it to use two different loops.  
It will first go through the top packages, and then the sub-packages.
I also swapped the symbol for uniqueness to _only_ use the name of the package, and not both name and version, so that it will not add a package with two different versions.

I updated the test cases with new packages which I installed via a similar method of the one that was used in the other package managers (and included comments for reproducibility).

New code will be pushed when I'm done with the other changes you have asked for :)

Reply:
I'm feeling like all we have to do is parse the top packages. The sub packages should exist on top level as well. If there are some version constraints in the sub packages, NuGet resolves it and pick up one version. But I'm not familiar with NuGet. If you think we need to parse the sub packages, could you tell me the document for it?

Reply:
Looking through the "complex" json file, I actually think you are correct, it seems like all sub-packages are in the file as `Transitive` dependencies in contrary to the ones installed by the user, which are `Direct`... I will double check to make sure that this is correct and if so, just drop the second pass of package traversal!

Reply:
From what I can see in the files I've tested with, you are correct!
So shortened the code a bit and updated the tests :)

Reply:
>it seems like all sub-packages are in the file as Transitive dependencies in contrary to the ones installed by the user, which are Direct...

Yes, yes, that is exactly I wanted to say. If we pick versions in the sub packages, it would be false positive because v9.0.1 is actually not installed in the example I shared.

Thanks for your confirmation!